### PR TITLE
Add /system/info endpoint for faas-swarm

### LINKED ISSUE #16: Support /system/info endpoint
We need to support the /system/info endpoint which has been updated in the latest release of faas-provider

## Tasks:

- Update vendored package for https://github.com/openfaas/faas-provider to 0.6
- Add new handler for system info endpoint included in faas-provider package
- Inject build version and SHA into binary at build-time (look at faas-cli as an example of how to do this)
- Add unit test for the HTTP handler

### PR DESCRIPTION
**Pull request by viveksyngh**
The purpose of this change is to make faas-swarm compatible with the
latest version of the API Gateway in OpenFaaS. Each provider should
be able to advertise the underlying provider and the version of the
code.

A new version of faas-provider (0.6) is vendored into the repo
which brings in an optional health endpoint (turned off by default).

The version and SHA is being passed to the provider from server.go.
Dockerfile is updated to inject version and SHA string at build time in
version.go.

A unit test has been added for the /system/info endpoint.

Signed-off-by: Vivek Singh <vivekkmr45@yahoo.in>

Fixes #16 

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by viveksyngh** [comment]
Test result for faas-swarm:
tested on Docker for Mac

```
ID                  NAME                MODE                REPLICAS            IMAGE                              PORTS
ghi8yww0cfr1        func_alertmanager   replicated          1/1                 prom/alertmanager:v0.7.1
5mgpou111zsh        func_base64         replicated          1/1                 functions/alpine:latest
yaz43m9d6xxu        func_echoit         replicated          1/1                 functions/alpine:latest
1l1ybyufyfvn        func_faas-swarm     replicated          1/1                 viveksyngh/faas-swarm:0.3.0        *:8081->8080/tcp
mv57f12qgq4l        func_gateway        replicated          1/1                 functions/gateway:0.7.3            *:8080->8080/tcp
zbtb8kpvus4p        func_hubstats       replicated          1/1                 functions/hubstats:latest
enrvvc5xn38v        func_markdown       replicated          1/1                 functions/markdown-render:latest
lzlkzo19c40k        func_nats           replicated          1/1                 nats-streaming:0.6.0
r83sp468soyg        func_nodeinfo       replicated          1/1                 functions/nodeinfo:latest
en36uzdi43bg        func_prometheus     replicated          1/1                 prom/prometheus:v2.0.0             *:9090->9090/tcp
rtzqull0ou1e        func_queue-worker   replicated          1/1                 functions/queue-worker:0.4
ckonb7npm55z        func_wordcount      replicated          1/1                 functions/alpine:latest
```

```
curl http://localhost:8081/system/info | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   123  100   123    0     0  17998      0 --:--:-- --:--:-- --:--:-- 20500
{
   "orchestration" : "swarm",
   "provider" : "faas-swarm",
   "version" : {
      "release" : "",
      "sha" : "18ee2f8c90a09104a987ae51f8693e4d572a45c0"
   }
}
```

**Comment by alexellis** [comment]
@johnmccabe should we look up the last release number from `git` when we're building outside of a Travis release?

This would give something like `0.2.4-dirty` or `0.2.4-dev`?

File: test/info_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	if required := http.StatusOK; rr.Code != required {
+		t.Errorf("handler returned wrong status code: got %v, but wanted: %v", rr.Code, required)
+	}
+
+	if infoRequest.Orchestration != handlers.SwarmIdentifier {
+		t.Errorf("handler returned wrong orchestration: got %v, but wanted: %v", infoRequest.Orchestration, handlers.SwarmIdentifier)
+	}
+
+	if infoRequest.Provider != handlers.SwarmProvider {
+		t.Errorf("handler returned wrong provider: got %v, but wanter: %v", infoRequest.Provider, handlers.SwarmProvider)
```
Comment (alexellis) [inline test/info_test.go]: There is a typo on this line `wanted`

Can we use: `etc.... - want: %v, got: %v` instead please?

File: handlers/info.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package handlers
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/openfaas/faas-provider/types"
+)
+
+//Identifier and Provider strings for swarm provider
```
Comment (alexellis) [inline handlers/info.go]: I think the comment should be on the const itself.

**Comment by viveksyngh** [comment]
Sure I will do these changes. I need help with testing `armhf` image for this. Image is getting built successfully but not able to deploy it. Anything I am missing in Dockerfile?

**Comment by johnmccabe** [comment]
@alexellis I'd be inclined to just use `git describe --tags`, would give you the last tag, commits since that tag and the current sha - `0.6.4-4-ge03d816` and mark that dirty.

**Comment by alexellis** [comment]
@viveksyngh what error do you get when you deploy on Scaleway?

**Comment by alexellis** [comment]
@johnmccabe would you be able to provide @viveksyngh with the equivalent command to put in his Dockerfile or is it better as a follow-up PR from yourself?

**Comment by viveksyngh** [comment]
@alexellis I don't get any error. Service does not come up.  

```
ID                  NAME                MODE                REPLICAS            IMAGE                                   PORTS
shgnsdqry25t        func_alertmanager   replicated          1/1                 alexellis2/alertmanager-armhf:0.5.1
wcneaklsmkc1        func_echoit         replicated          1/1                 functions/alpine:latest-armhf
pzkbereudssx        func_faas-swarm     replicated          0/1                 viveksyngh/faas-swarm:0.1.5-armhf       *:8081->8080/tcp
mm4vzmss3n9i        func_gateway        replicated          1/1                 functions/gateway:0.7.0-armhf           *:8080->8080/tcp
wig81sqw5gq8        func_markdown       replicated          1/1                 functions/markdownrender:latest-armhf
0hqthrv971f8        func_nats           replicated          1/1                 nats-streaming:0.6.0
6s4nkoaepxd0        func_nodeinfo       replicated          1/1                 functions/nodeinfo:latest-armhf
m37yyfxc76h8        func_prometheus     replicated          1/1                 alexellis2/prometheus:2.0-armhf         *:9090->9090/tcp
0wehg8sfzsyy        func_queue-worker   replicated          1/1                 functions/queue-worker:0.4-armhf
v9ngwjpynz0p        func_wordcount      replicated          1/1                 functions/alpine:latest-armhf
```

**Comment by alexellis** [comment]
Try this:

`docker service ps --no-trunc func_faas-swarm` or `docker service logs ....`

File: Dockerfile [hunk 3]
```diff
@@ -8,14 +8,21 @@ WORKDIR /go/src/github.com/openfaas/faas-swarm
...
+COPY version    version
+COPY .git       .git
 COPY server.go  .
 
 RUN curl -sL https://github.com/alexellis/license-check/releases/download/0.2.2/license-check > /usr/bin/license-check \
     && chmod +x /usr/bin/license-check
 RUN license-check -path ./ --verbose=false "Alex Ellis" "OpenFaaS Project"
 
 RUN gofmt -l -d $(find . -type f -name '*.go' -not -path "./vendor/*") \
-  && CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o faas-swarm .
```
Comment (alexellis) [inline Dockerfile]: You shouldn't have deleted this line as it may have changed the name of the output binary.

File: Dockerfile [hunk 4]
```diff
@@ -8,14 +8,21 @@ WORKDIR /go/src/github.com/openfaas/faas-swarm
 COPY vendor     vendor
 COPY handlers	  handlers
 COPY types      types
+COPY version    version
+COPY .git       .git
```
Comment (alexellis) [inline Dockerfile]: @johnmccabe is this needed? Is it avoided in the CLI due to the way we `COPY .`?

**Comment by viveksyngh** [comment]
I am getting this error .. `no suitable node (unsupported platform on 1 node)`

File: Dockerfile (hunk 4 above)
Comment (viveksyngh) [reply Dockerfile]: In CLI we do `copy .  .`  I did not copy everything because it will may increase the size of  image

Comment (johnmccabe) [reply Dockerfile]: Theres no real need to be selective surely? Its a multi-stage build so the resulting image is `scratch` + the binary.

Comment (viveksyngh) [reply Dockerfile]: Sure .. I will change that/

**Comment by alexellis** [comment]
My thinking was this: can we run the git commands outside of Docker in the makefile? Then we don't need to copy the git directory in. I think to get this work done what you have now is probably fine.

**Comment by viveksyngh** [comment]
@alexellis Seems like there is a mismatch between docker platform/os and `faas-swarm` service platform/os

`docker` : 
```
OSType: linux
Architecture: armv7l
```

```
"Placement": {
                    "Constraints": [
                        "node.role == manager",
                        "node.platform.os == linux"
                    ],
                    "Platforms": [
                        {
                            "Architecture": "amd64",
                            "OS": "linux"
                        }
                    ]
                },
```

I may need help of testing `faas-swarm` armhf image.

**Comment by viveksyngh** [comment]
Tested on `arm` as well. 

```
ID                  NAME                MODE                REPLICAS            IMAGE                                   PORTS
q6nrimiyrbjw        func_alertmanager   replicated          1/1                 alexellis2/alertmanager-armhf:0.5.1
azqirqeq976r        func_echoit         replicated          1/1                 functions/alpine:latest-armhf
qb3ignx6chix        func_faas-swarm     replicated          1/1                 functions/faas-swarm:latest-dev         *:8081->8080/tcp
sog48c0e53sb        func_gateway        replicated          1/1                 functions/gateway:0.7.0-armhf           *:8080->8080/tcp
bxadxxpo7hle        func_markdown       replicated          1/1                 functions/markdownrender:latest-armhf
nr4fohnogwrp        func_nats           replicated          1/1                 nats-streaming:0.6.0
qb8o451vny7a        func_nodeinfo       replicated          1/1                 functions/nodeinfo:latest-armhf
yzpevb556npo        func_prometheus     replicated          1/1                 alexellis2/prometheus:2.0-armhf         *:9090->9090/tcp
sk28yu2ksnk2        func_queue-worker   replicated          1/1                 functions/queue-worker:0.4-armhf
jlly4s7bbwfu        func_wordcount      replicated          1/1                 functions/alpine:latest-armhf
```

```
curl http://95b6c0b9-a961-42db-8a1d-7981328139d1.pub.cloud.scaleway.com:8081/system/info | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   123  100   123    0     0    227      0 --:--:-- --:--:-- --:--:--   227
{
   "version" : {
      "release" : "",
      "sha" : "7179e91598a57da21aa35a03499a8709f140b77a"
   },
   "orchestration" : "swarm",
   "provider" : "faas-swarm"
}```

File: Dockerfile (hunk 3 above)
Comment (viveksyngh) [reply Dockerfile]: I have edited it to inject the version and commit id.

**Comment by johnmccabe** [comment]
@alexellis re: the `.git` dir, it's used in the builder, the alternative would be to set a build arg which seems like overkill

**Comment by alexellis** [comment]
OK so this is the kind of output we can get from the command John suggested

```
$ git describe --tags
0.7.7-1-ge442da3
```

It'd be ideal if we just had the version without the extra data which we currently take in via the SHA anyway.

Should we just use `cut` - or is there another `git` command that gives only the latest tag @johnmccabe ?

**Comment by alexellis** [comment]
This also seems to work for getting the tag - `git tag --sort=-committerdate|head -n 1`

**Comment by viveksyngh** [comment]
@alexellis These commands will display the last release of the provider. Should I update this PR to use `git tag --sort=-committerdate|head -n 1`?

**Comment by johnmccabe** [comment]
You can do something like this..
```
git describe --all --exact-match `git rev-parse HEAD` | grep tags | sed 's/tags\///
```
Has a tag if its from a tagged SHA, empty if not, then in the version logic print dirty, as long as you're still printing the sha in the output/response you'd good - tag/sha, dirty/sha.

**Comment by viveksyngh** [comment]
Makes sense. Should it be `dirty` or `dev` is fine?

**Comment by johnmccabe** [comment]
personally I like `dev`, its clearer

**Comment by viveksyngh** [comment]
@alexellis @johnmccabe  I have updated the PR with requested changes and also updated the test result.

**Comment by alexellis** [comment]
I'm OK to merge this change - I have tested it and got this result:

```json
{"provider":"faas-swarm","version":{"sha":"050ae76e5d5bfceb760ab144cd341b07f4d56f77",
"release":"dev"},"orchestration":"swarm"}
```

What I really wanted was this:

```json
{"provider":"faas-swarm","version":{"sha":"050ae76e5d5bfceb760ab144cd341b07f4d56f77",
"release":"0.4.2-dev"},"orchestration":"swarm"}
```

Let's fix up in a follow-up PR so that we can start integrating.

**Pull request merged**